File: codex_poting/java/src/main/java/com/example/demo/mybatis/validator/BookDataValidatorMybatis.java
package com.example.demo.mybatis.validator;

import com.example.demo.exception.ForeignKeyReferenceNotFoundException;
import com.example.demo.exception.UniqueConstraintValidationException;
import com.example.demo.mybatis.generator.entity.BookEntity;
import com.example.demo.mybatis.generator.entity.BookEntityExample;
import com.example.demo.mybatis.generator.entity.BookGenreEntity;
import com.example.demo.mybatis.generator.entity.PublisherEntity;
import com.example.demo.mybatis.generator.mapper.BookMapper;
import com.example.demo.mybatis.generator.mapper.BookGenreMapper;
import com.example.demo.mybatis.generator.mapper.PublisherMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Profile;
import org.springframework.orm.ObjectOptimisticLockingFailureException;
import org.springframework.stereotype.Component;

import java.util.Objects;

@Component
@Profile("mybatis")
@RequiredArgsConstructor
public class BookDataValidatorMybatis {
    private final BookMapper bookMapper;
    private final PublisherMapper publisherMapper;
    private final BookGenreMapper bookGenreMapper;

    public void foreignKeyValidate(Long publisherId, Long genreId) {
        final var publisher = publisherMapper.selectByPrimaryKey(publisherId);
        if (Objects.isNull(publisher)) {
            throw new ForeignKeyReferenceNotFoundException(PublisherEntity.class, publisherId);
        }

        final var bookGenre = bookGenreMapper.selectByPrimaryKey(genreId);
        if (Objects.isNull(bookGenre)) {
            throw new ForeignKeyReferenceNotFoundException(BookGenreEntity.class, genreId);
        }
    }

    public void versionValidate(BookEntity book, Long requestVersion) {
        if (!Objects.equals(book.getVersion(), requestVersion)) {
            throw new ObjectOptimisticLockingFailureException(BookEntity.class, book.getId());
        }
    }

    public void uniqueIsbnValidate(String isbn, Long bookId) {
        final var example = new BookEntityExample();
        example.createCriteria().andIsbnEqualTo(isbn);
        bookMapper.selectByExample(example).stream()
            .filter(book -> !Objects.equals(book.getId(), bookId))
            .findFirst()
            .ifPresent(book -> {
                throw new UniqueConstraintValidationException("book", "isbn", isbn);
            });
    }
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/mybatis/validator/BookDataValidatorMybatis.kt
package com.example.demo.mybatis.validator

import com.example.demo.exception.ForeignKeyReferenceNotFoundException
import com.example.demo.exception.UniqueConstraintValidationException
import com.example.demo.mybatis.generator.entity.BookEntity
import com.example.demo.mybatis.generator.entity.BookEntityExample
import com.example.demo.mybatis.generator.entity.BookGenreEntity
import com.example.demo.mybatis.generator.entity.PublisherEntity
import com.example.demo.mybatis.generator.mapper.BookGenreMapper
import com.example.demo.mybatis.generator.mapper.BookMapper
import com.example.demo.mybatis.generator.mapper.PublisherMapper
import org.springframework.context.annotation.Profile
import org.springframework.orm.ObjectOptimisticLockingFailureException
import org.springframework.stereotype.Component
import java.util.*
import java.util.function.Consumer

@Component
@Profile("mybatis")
class BookDataValidatorMybatis(
    private val bookMapper: BookMapper,
    private val publisherMapper: PublisherMapper,
    private val bookGenreMapper: BookGenreMapper
) {
    fun foreignKeyValidate(publisherId: Long?, genreId: Long?) {
        val publisher = publisherMapper.selectByPrimaryKey(publisherId)
        if (Objects.isNull(publisher)) {
            throw ForeignKeyReferenceNotFoundException(PublisherEntity::class.java, publisherId)
        }
        val bookGenre = bookGenreMapper.selectByPrimaryKey(genreId)
        if (Objects.isNull(bookGenre)) {
            throw ForeignKeyReferenceNotFoundException(BookGenreEntity::class.java, genreId)
        }
    }

    fun versionValidate(book: BookEntity, requestVersion: Long?) {
        if (book.getVersion() != requestVersion) {
            throw ObjectOptimisticLockingFailureException(BookEntity::class.java, book.getId())
        }
    }

    fun uniqueIsbnValidate(isbn: String?, bookId: Long?) {
        val example = BookEntityExample()
        example.createCriteria().andIsbnEqualTo(isbn)
        bookMapper.selectByExample(example).stream().filter { book: BookEntity? -> book!!.getId() != bookId }
            .findFirst().ifPresent(
                Consumer { book: BookEntity? ->
                    throw UniqueConstraintValidationException("book", "isbn", isbn)
                })
    }
}
